### ISSUE TITLE
n_classes/clusters and latent_dim

### ISSUE THREAD

**Issue by guenthereder** [xuyxu/Deep-Clustering-Network#1]
The dimension of the output of the encoder should be 'latent_dim' but is currently 'n_classes' (=n_clusters) guarded by following assertion:
autoencoder.py - line 20
assert self.dims_list[self.n_layers // 2] == self.n_classes

The k_means centers are initialized correctly in kmeans.py - line 20
self.clusters = np.zeros((self.n_clusters, self.n_features))

However, then overwritten in 'def init_cluster(self, X, indices=None)' where X arrives with a 'wrong' dimension.
Line 36-37 then fit and assigned new k_means centers with dimension of n_clusters x n_clusters.

It seems to suffice to change the encoder output to 'latent_dim' (and adjust the assertion).
Thanks a lot for the effort of implementing DCN in pytorch!

**Comment by xuyxu** [comment]
Hi @guenthereder, thanks a lot!

I will work on your suggestions soon and test the performance ;-)

**Comment by xuyxu** [comment]
Hi @guenthereder, once again, thanks for your kind reminders. I would like to ask that whether you can reproduce the results on MNIST reported in the original paper.

To be honest, I cannot achieve a  good performance just like those reported in the original paper, and it looks like their official implementations also have this problem: https://github.com/boyangumn/DCN-New/issues/2

**Comment by guenthereder** [comment]
@xuyxu, ~~no I can not reproduce the results claimed in the paper~~ Yes I can confirm the claims in the paper. I tried some parameter variations and posted the results here: https://github.com/guenthereder/Deep-Clustering-Network (a fork of your implementation).
Regardless, thanks again for implementing DCN, saved me a lot of time.

**Comment by guenthereder** [comment]
@xuyxu I found a small mistake, you should shuffle the training data, less so for the testing data: mnist.py - line 102.

**Comment by xuyxu** [comment]
I am sorry for my late response @guenthereder. I will correct them, thanks again ;-)

**EDIT:** I have checked your forked repo. It looks like that you have replaced the k-means with mean shift, right? If so, can I add an acknowledgement on my `README.md` so that other people can also know that you are more close to the results in the paper?

**Comment by guenthereder** [comment]
Using mean shift is not (fully) implemented yet, by default my fork uses k-means.
Acknowledgement: Yes sure, thanks.

**Comment by xuyxu** [comment]
Glad to hear that @guenthereder, happy hacking!
